osu.Game/Screens/Select/FilterQueryParser.cs

```diff
@@ -312,11 +322,43 @@ private static bool tryUpdateCriteriaRange<T>(ref FilterCriteria.OptionalRange<T
 
         private static bool tryUpdateLengthRange(FilterCriteria criteria, Operator op, string val)
         {
-            if (!tryParseDoubleWithPoint(val.TrimEnd('m', 's', 'h'), out double length))
+            List<string> parts = new List<string>();
+
+            GroupCollection match = null;
+            match ??= tryMatchRegex(val, @"^((?<hours>\d+):)?(?<minutes>\d+):(?<seconds>\d+)$");
+            match ??= tryMatchRegex(val, @"^((?<hours>\d+(\.\d+)?)h)?((?<minutes>\d+(\.\d+)?)m)?((?<seconds>\d+(\.\d+)?)s)?$");
+            match ??= tryMatchRegex(val, @"^(?<seconds>\d+(\.\d+)?)$");
+
+            if (match == null)
                 return false;
 
-            int scale = getLengthScale(val);
-            return tryUpdateCriteriaRange(ref criteria.Length, op, length * scale, scale / 2.0);
+            if (match["seconds"].Success)
```

Comment:
Reassembling is necessary unless there is a better way to do all checks without assembling these match results to an ordered sequence.

The checks which rely on the array:
- All of them should be less than 60 except the last part
- All of them shouldn't contain a decimal dot except the first part
- Determine the minimum scale